File: c/src/tree_sort.c
#include "../sorting_algorithms.h"
#include <stdlib.h>

typedef struct TreeNode {
    int value;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

static TreeNode *insert(TreeNode *root, int value) {
    if (!root) {
        TreeNode *node = (TreeNode *)malloc(sizeof(TreeNode));
        if (!node) return NULL;
        node->value = value;
        node->left = node->right = NULL;
        return node;
    }
    if (value < root->value) root->left = insert(root->left, value);
    else root->right = insert(root->right, value);
    return root;
}

static void inorder(TreeNode *root, int *arr, size_t *idx) {
    if (!root) return;
    inorder(root->left, arr, idx);
    arr[(*idx)++] = root->value;
    inorder(root->right, arr, idx);
}

static void free_tree(TreeNode *root) {
    if (!root) return;
    free_tree(root->left);
    free_tree(root->right);
    free(root);
}

void tree_sort(int *arr, size_t n) {
    TreeNode *root = NULL;
    for (size_t i = 0; i < n; ++i) root = insert(root, arr[i]);
    size_t idx = 0;
    inorder(root, arr, &idx);
    free_tree(root);
}

File: python/sorting_algorithms/algos/tree_sort.py
from typing import Optional, Sequence
from ..base import _setup_tracer, _finalize, SortTracer, Number


class _Node:
    __slots__ = ("value", "left", "right")

    def __init__(self, value: Number):
        self.value = value
        self.left: Optional["_Node"] = None
        self.right: Optional["_Node"] = None


def tree_sort(data: Sequence[Number], *, trace: bool = False, tracer: Optional[SortTracer] = None):
    arr = list(data)
    tracer, ret_trace = _setup_tracer("tree_sort", trace, tracer)
    if not arr:
        return _finalize(arr, tracer, ret_trace)

    root = _Node(arr[0])
    if tracer:
        tracer.record([root.value], (), "root")

    def insert(node: _Node, value: Number):
        if value <= node.value:
            if node.left is None:
                node.left = _Node(value)
            else:
                insert(node.left, value)
        else:
            if node.right is None:
                node.right = _Node(value)
            else:
                insert(node.right, value)

    for v in arr[1:]:
        insert(root, v)
        if tracer:
            tracer.record(arr, (), f"insert {v}")

    output: list[Number] = []

    def inorder(node: Optional[_Node]):
        if node is None:
            return
        inorder(node.left)
        output.append(node.value)
        if tracer:
            tracer.record(output + [node.value], (), "inorder append")
        inorder(node.right)

    inorder(root)
    return _finalize(output, tracer, ret_trace)
